app/services/katello/candlepin_event_listener.rb

```diff
@@ -0,0 +1,104 @@
+module Katello
+  class CandlepinEventListener
+    PROCESSED_COUNT_CACHE_KEY = 'candlepin_events_processed'.freeze
+    FAILED_COUNT_CACHE_KEY = 'candlepin_events_failed'.freeze
+    AMQP_QUEUE_NAME = 'event.org.candlepin.audit.AMQPBusPublisher'.freeze
+
+    CandlepinEvent = Struct.new(:message_id, :subject, :content)
+
+    @logger = ::Foreman::Logging.logger('katello/candlepin_events')
+    @failed_count = 0
+    @processed_count = 0
+
+    def self.start_service
+      loop do
+        begin
+          result = Katello::CandlepinListeningService.instance.start
+
+          break if result == :connected
+
+          @logger.info("Attempting to restart Candlepin Listening Service")
+          sleep 5
+        end
+      end
+    end
+
+    def self.run
+      @thread.kill if @thread
+
+      # run in own thread so connecting to qpid won't block the main process
+      @thread = Thread.new do
+        initialize_listening_service
+        start_service
+
+        ActiveRecord::Base.connection_pool.with_connection do
```

Comment:
@jturel Revisiting the `Rails.application.executor.wrap` method, I was looking at [the Rails docs on threading](https://guides.rubyonrails.org/threading_and_code_execution.html) and this part jumped out at me:

> Prior to Rails 5.0, some of these were handled by separate Rack middleware classes (such as ActiveRecord::ConnectionAdapters::ConnectionManagement), or directly wrapping code with methods like ActiveRecord::Base.connection_pool.with_connection. The Executor replaces these with a single more abstract interface.

It seems like they are suggesting to use either ` `Rails.application.executor.wrap`  or `
Rails.application.reloader.wrap` instead of `ActiveRecord::Base.connection_pool.with_connection` for Rails 5 apps. We should probably stick to convention Rails specifies here, but maybe you saw some reason to use the `with_connection` method over the application wrappers. Any thoughts?

Reply:
I didn't realize that was handled by the executor! When I was researching I just came across with_connection. The application.executor should work, but just want to mention that all that needs to be considered here (I think) is the database connection since the reloading is handled in engine.rb So, some preciseness is lost in the intention of the call, but the documentation does seem to indicate using the executor is preferred.

Reply:
Updated to use Rails.application.executor.wrap